src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.CodeFixes/BinaryFormatterUnsafeDeserializeFixer.cs

```diff
@@ -0,0 +1,94 @@
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.Analyzers;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.CodeFixes
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = "UA0012 CodeFix Provider")]
+    public sealed class BinaryFormatterUnsafeDeserializeFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(BinaryFormatterUnsafeDeserializeAnalyzer.DiagnosticId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            if (root is null)
+            {
+                return;
+            }
+
+            var node = root.FindNode(context.Span, false, true);
+
+            if (node is null)
+            {
+                return;
+            }
+
+            // Register a code action that will invoke the fix.
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    CodeFixResources.BinaryFormatterUnsafeDeserializeTitle,
+                    cancellationToken => ReplaceUnsafeDeserializationAsync(context.Document, node, cancellationToken),
+                    nameof(CodeFixResources.BinaryFormatterUnsafeDeserializeTitle)),
+                context.Diagnostics);
+        }
+
+        private static async Task<Solution> ReplaceUnsafeDeserializationAsync(Document document, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            var project = document.Project;
+            var slnEditor = new SolutionEditor(project.Solution);
+            var docEditor = await slnEditor.GetDocumentEditorAsync(document.Id, cancellationToken).ConfigureAwait(false);
+            var docRoot = (CompilationUnitSyntax)docEditor.OriginalRoot;
+
+            if (!(node is MemberAccessExpressionSyntax))
+            {
+                // stop processing - the code fixer only fixes member invocations of UnsafeDeserialize
+                return document.Project.Solution;
+            }
+
+            var memberExpression = node as MemberAccessExpressionSyntax;
+
+            if (memberExpression!.Expression is IdentifierNameSyntax)
+            {
```

Comment:
Many of these actions should be able to be done in a language agnostic way via the DocumentEditor and the SyntaxGenerator it produces